[Dallas] Fix timings for multiple tasks running Dallas sensors

TD-er opened the pull request:
When running multiple instances of a Dallas plugin, it was possible a sensor was read before it was finished measuring the temperature.

TD-er committed:
[Dallas] Fix timings for multiple tasks running Dallas sensors

When running multiple instances of a Dallas plugin, it was possible a sensor was read before it was finished measuring the temperature.

TD-er committed:
[Dallas] Move redundant code from P004, P080 and P100 to a Dallas helper

TD-er committed:
[Dallas] Cleanup P004 DS18b20 plugin code

TD-er committed:
[Dallas] Reschedule at normal interval to prevent drift caused by reading

Reading a sensor value from a Dallas sensor may take upto 800 msec, based on the set resolution.
This means the set interval for producing a sample will drift by roughly 800 - 900 msec per reading.
By recording the time when starting the reading and using that to reschedule the task timer this drift is minimized.

TD-er commented:
OK, this PR does include the long overdue TLC the Dallas plugin needed.
I believe it is running quite stable and predictable now on my test system.

The code duplication between P004, P080 and P100 is now also a lot less.
The main fixes for P004 is about making sure a sensor is only read after its measurement period and also preventing interval drifting caused by rescheduling the task with set interval after the measurement is done.  (still not perfect)

I would love to see some real world testing.
@uzi18 @thomastech @tonhuisman Please be my guest.
I will make a test build, but I also believe you are capable of building your own :)

TD-er commented:
Test build is ready:
- [ESPEasy_ESP32_mega-20201102-21-PR_3367.zip](https://www.dropbox.com/s/gvkqxaxx1bzk7x4/ESPEasy_ESP32_mega-20201102-21-PR_3367.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20201102-21-PR_3367.zip](https://www.dropbox.com/s/tztemqbhfik68hm/ESPEasy_ESP82xx_mega-20201102-21-PR_3367.zip?dl=0)

thomastech commented:
> I would love to see some real world testing.
@uzi18 @thomastech @tonhuisman Please be my guest.

I don't have a spare "testing" node with a Dallas sensor on it. But I'll build one and do a very basic go / no-go validation for you. Could be today, or could be later this week.

FWIW, my existing Dallas nodes are in inconvenient places. One is in the attic and the other is under the house (crawl space). So I don't use them to validate firmware since recovery from bad code is messy and inconvenient. Fortunately since early 2020 they have been very reliable (no reboots in months), which is due to your stability fixes. Thank you!

- Thomas

thomastech commented:
Tested ESP_Easy_mega_20201110_test_ESP8266_4M1M_VCC on a Wemos D1 Mini. 
- One DS18B20 with 4.7K pullup using GPIO14 (D5).
- One second interval.
- Verified 9, 10, 11, 12 bit modes.
- Temperature value shown in device task list is valid.
- No other tests performed.

**Edit:** I recommend that your PR also include an update to the P004 plugin's readthedocs link. Because there's two rst files, one with little helpful info and one that is useful. 

That is to say, the P004.rst file is a basic template with little info. But there's in-depth info in P004_DS18b20.rst. I suggest that P004.rst include P004_DS18b20.rst without requiring that the reader click another link. And while at it, greatly reduce the size of the first photo in P004_DS18b20.rst. Just my simple-minded recommendations.

\- Thomas

tonhuisman commented:
> I would love to see some real world testing.
> @uzi18 @thomastech @tonhuisman Please be my guest.

Sure.
Just updated my production-ESP-01 that measures both the Deep-freezer and Fridge temperatures using 2 DS18B20's in parallel. First impression: It looks to be running just fine, but I'll keep an eye on it for the next few days, though I don't _expect_ things to go hay-wire 😛

TD-er commented:
> though I don't _expect_ things to go hay-wire

Nope, only 1-wire :)

TD-er committed:
[Dallas] Improve timings to comply with datasheet timings

TD-er committed:
[Dallas] Fix build issue on ESP32

TD-er commented:
Made a new test build including the improved timings.
- [ESPEasy_ESP32_mega-20201102-23-PR_3367.zip](https://www.dropbox.com/s/8pah5spdj9drlf3/ESPEasy_ESP32_mega-20201102-23-PR_3367.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20201102-23-PR_3367.zip](https://www.dropbox.com/s/9v2arfgtkizqwck/ESPEasy_ESP82xx_mega-20201102-23-PR_3367.zip?dl=0)

By the way, this entire PR could not have been possible to this extend if it wasn't for my new toy... ehh tool, my new Saleae logic analyzer.
It isn't sponsored by them, but it is bought from the Patreon money.
Just some background info of a happy user :)

thomastech commented:
I flashed ESP_Easy_mega_20201111_test_ESP8266_4M1M_VCC. Appears the same, the results match my earlier test.

\-Thomas

TD-er commented:
Thanks, so at least nothing got broken :)
I also ordered a bag full of the cheapest "Dallas" sensors I could find, to be able to test with ones of disputable origin.
The ones I have here (3 pieces, ordered years ago) do all work together just fine, but it is also the "magic" threshold of number of clones that may work together.

TD-er committed:
[Docs] Make the P004 Dallas documentation a bit more informative

TD-er commented:
I just added the task index to the pull down combobox, to help decide which address should be used.
![image](https://user-images.githubusercontent.com/3751318/98807752-69a3ee00-241b-11eb-9b29-055c24f4439e.png)

I can also add the task name and/or variable name
Is this clear enough?

Here have been requests to name the sensors, but not sure how to do that other then taskname/variable name.

tonhuisman commented:
I'd go for the number only, to avoid that you need to leave the name out later for the 'limit build size' builds... or just do that now already, and show both the tasknumber and name for the not size-limited builds.

TD-er commented:
The build doesn't increase in size, as I read the name from the ExtraTaskSettings which is active in memory anyway when searching for other Dallas tasks.

For example:
![image](https://user-images.githubusercontent.com/3751318/98809269-92c57e00-241d-11eb-98c3-a429fdd11371.png)

(this one has a 1-off bug in the task nr, but that's already fixed...)

tonhuisman commented:
> The build doesn't increase in size

That's even better, then just leave as you did, looks very usable 👍

TD-er commented:
Just some teaser to show what I'm working on:
![image](https://user-images.githubusercontent.com/3751318/98840615-4c851480-2447-11eb-878c-674d7d277b0a.png)

tonhuisman commented:
How to get the second temperature to Domoticz using the http or mqtt controller? A second instance will be the easiest way I guess, don't think Domoticz has a double temperate device.
But if you're short on free device slots, this is helpful.

TD-er commented:
> How to get the second temperature to Domoticz using the http or mqtt controller? A second instance will be the easiest way I guess, don't think Domoticz has a double temperate device.
> But if you're short on free device slots, this is helpful.

Therefore it is optional, with the default set to "Single".
The warning also suggests Domoticz would not be happy when you change it.

thomastech reviewed (changes requested):
Thanks for updating the DS18b20 docs. 

I made some minor edits to P004_DS18b20.rst (text improvements, friendly image sizes). You are welcome to add the revised file to your PR:
[P004_DS18b20.zip](https://github.com/letscontrolit/ESPEasy/files/5526134/P004_DS18b20.zip)

My edits do not include the last minute changes you made to the P004 plugin. So feel free to further revise the docs.

\- Thomas

TD-er commented:
Thanks I will include it.

Misiu commented:
> Here have been requests to name the sensors, but not sure how to do that other then taskname/variable name.

@TD-er would it be possible to show the device name that the sensor is assigned to?
The use case is very simple: I want to connect four sensors, typically you create a device per sensor. If you assign the first DS18B20 to a device (for example temperature1), when you define a second device and click on address dropdown you should see four positions (because you have 4 sensors):
11-22-33-44-55-66-77-88
12-22-33-44-55-66-77-88 (temperature1)
13-22-33-44-55-66-77-88
14-22-33-44-55-66-77-88

This way you won't select the address that is already in use.
Is something like this possible?

BTW:
-is the 4 DS18B20 limit still valid (https://github.com/letscontrolit/ESPEasy/issues/1084#issuecomment-512292521)?
-can we rescan for new devices (https://github.com/letscontrolit/ESPEasy/issues/1085#issuecomment-482051441)? Idea is to connect DS18B20 one by one .I have sensors on long wires and it is hard to tell which one is which if I connect all of them at once. Current workaround is to configure first, power-off, plug second, power-on, configure and so one. A tiny button near the address dropdown that triggers rescan would be awesome!

TD-er committed:
[Dallas] Support multiple temp sensors in 1 task for simultaneous reads

TD-er commented:
Is this what you mean?
![image](https://user-images.githubusercontent.com/3751318/98873659-152f5b80-2479-11eb-9271-60468fab4317.png)

About rescanning... just opening the page for the task will perform a scan.

I will now make a new test build so you can also play around to see if it is already as you imagined it.
The newly added feature to have upto 4 sensors in one task also allows to read the sensors at (almost) the same time.

See here the logic analyzer output of sampling 3 sensors from one task:
![image](https://user-images.githubusercontent.com/3751318/98873989-adc5db80-2479-11eb-9ef8-bc5c6bc37510.png)

It first sends 3 conversion start commands to the sensors, waits 800 msec (starting right after the first sensor starts measuring) and reads 3 sensor values.
So the measuring periods overlap, which is very hard to do when running from separate tasks.
4 sensors at 12 bit can be read like this is just under a second. So you could even set the task interval to 1 second and read upto 4 sensors at the same time almost continously.


Sending of 3 start commands:
![image](https://user-images.githubusercontent.com/3751318/98874535-ba96ff00-247a-11eb-8f89-4efa56fbd096.png)

800 msec after the first start command was sent, the 3 read commands:

![image](https://user-images.githubusercontent.com/3751318/98874637-e87c4380-247a-11eb-91a9-7a73f43c6d92.png)

And the accuracy of the timings:
Write 0: 60 usec down, 10 up:
![image](https://user-images.githubusercontent.com/3751318/98874711-13669780-247b-11eb-9cfe-5fa035e91de8.png)

Write 1: 6 usec down, 64 usec up:
![image](https://user-images.githubusercontent.com/3751318/98874786-35601a00-247b-11eb-900d-4799fe5359c4.png)

Read 0: 6 usec down, wait for 9 usec and see what the value is:
![image](https://user-images.githubusercontent.com/3751318/98874951-82dc8700-247b-11eb-94ce-18d0c16a5a54.png)

Reset: 480 usec down, line goes up and is pulled down for at least 60 usec to signal a presence:
![image](https://user-images.githubusercontent.com/3751318/98875051-bae3ca00-247b-11eb-88a8-74205dba30ab.png)

All values based on the documentation:
![image](https://user-images.githubusercontent.com/3751318/98875237-0f874500-247c-11eb-814a-c697d0053d2d.png)

Misiu commented:
@TD-er I have some spare Wemos D1 so I can give this a try 😊
I use a very old build, with devices instead of tasks (I think I missed the rename), so I think things are a bit different right now.
On my devices (running old firmware) I'm calculating the average from 10 readings and only then I'm sending the value via MQTT (per each sensor). I hope this will be possible with this update.

BTW the DS18B20 device limit is still 4?

TD-er commented:
Test build including the multiple devices per task feature:

- [ESPEasy_ESP32_mega-20201102-25-PR_3367.zip](https://www.dropbox.com/s/jz8j85fsx1edf5a/ESPEasy_ESP32_mega-20201102-25-PR_3367.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20201102-25-PR_3367.zip](https://www.dropbox.com/s/3972z0oemnhqkwh/ESPEasy_ESP82xx_mega-20201102-25-PR_3367.zip?dl=0)

TD-er commented:
> @TD-er I have some spare Wemos D1 so I can give this a try 😊
> I use a very old build, with devices instead of tasks (I think I missed the rename), so I think things are a bit different right now.
> On my devices (running old firmware) I'm calculating the average from 10 readings and only then I'm sending the value via MQTT (per each sensor). I hope this will be possible with this update.
> 
> BTW the DS18B20 device limit is still 4?

Please backup your settings first before trying this build.
The only limit for Dallas sensors I can think of is 4 per task and the maximum number of tasks you can have.

There is of course a practical limit of the number of sensors you can place on a single bus before exceeding the length limit (more like capacitance limit and max. power) of the cable.
When hitting this limit, you can try to use another GPIO pin (1 GPIO pin per task of max 4 Dallas sensors), but you can also use the same GPIO pin among several tasks running this Dallas temp sensor plugin.

To summarize:
- 1 GPIO per task running P004
- Multiple tasks running P004 may use the same GPIO (or another one)
- Max 4 Dallas temperature sensors per task
- resolution is set per task => thus also for all sensors in that same task running P004
- All sensors in the same task running P004 will be read at the same time (few usec offset to send the commands)

It is known that some fake Dallas sensors have a limit of 3 on the same bus.
You can try to group those (if applicable to you) in the same task and make a new task using a different GPIO for the next few.

TD-er committed:
[Dallas] Add statistics + speed up reading

Misiu commented:
@TD-er I flashed a new Wemos D1 with the test build.
Everything works fine, I can read from a single, two, and three sensors in a single Device 🚀 (I'll check more sensors in the evening)

I noticed a couple of things that could be improved in the UI (but I can live with them 🙂 )
1. If possible please change the name to **DS18B20**
![image](https://user-images.githubusercontent.com/1741838/98943226-7560e480-24ef-11eb-9f7c-5b35c4a3cbb1.png)

2. After selecting a new device type from the list you automatically go to Task settings, so the Close and Submit buttons don't make much sense. I think the user should confirm adding a new device. It's just a matter of removing
 `onchange="return dept_onchange(frmselect)"`
3. Why we are in Task Settings when the Tab is named Devices?
![image](https://user-images.githubusercontent.com/1741838/98944082-c91ffd80-24f0-11eb-857d-665bcab13863.png)
4. After changing the Output Data Type to Triple the Values default values don't update:
![image](https://user-images.githubusercontent.com/1741838/98944223-02f10400-24f1-11eb-9a1e-7c47935a84a5.png)
By default first value has the name `temperature` and 2 decimals, when having more than one others have an empty name and 0 decimals.
5. Clicking on `Delete` in Task (Device) properties navigates to `Add new device page`, but it should navigate to devices list.

Those are just small suggestions, they can be skipped.

I'll report back after some more tests.

P.S.
Is there a chance to tweak ESP Easy Flasher? Take a look at [tasmotizer](https://github.com/tasmota/tasmotizer)
Maybe the next version of Flasher could be built in a similar way? I have a 2yo son, so I can't help much, but I'll try to create something 🙂

TD-er commented:
Yep there are still some inconsistencies in the UI (and in the code)
Just some of the names we use in ESPEasy:

- Plugin: Piece of software interacting with a sensor.
- Task: An instance of a plugin. Can have 0 .. 3 controllers connected.
- Controller: Receives data from a task and sends it to an end point (e.g. some server)

About the defaults, that's indeed still something I need to do for this plugin.
Right now it is still a work in progress.
Will have a look at suggestions 1 ... 3. 

Suggestion 5 is something I don't agree on as you are still on the task editing page which has 2 states:
- Select the plugin
- Configure the task.

Misiu commented:
Thanks for the vocabulary description, I'll try to use the correct names from now on.
When you are in edit task view and you click on Delete you should go to `tasks list`, now you are redirected to `new task` view:
![espeasy_1](https://user-images.githubusercontent.com/1741838/98953308-d216cc00-24fc-11eb-8e04-ab2977cd5a63.gif)
I've recorded a GIF to show what I mean because I'm not sure if I described it correctly.

In many (if not most) applications I used (and I created) when the user deletes an item, he is redirected to the list of other items. I think this should work the same here.

Even if this is still WIP it already works great! Awesome work 🚀

Any thoughts about tasmotizer and similar software for ESP Easy?

TD-er commented:
I have not looked at Tasmotizer.
The flashing tool(s) really need some TLC too, as the ESPEasy flasher now does not support ESP32 and there are quite a lot of (cheap) USB to serial adapters that are not recognized by the ESPEasy flasher.

thomastech commented:
Feedback regarding change requests 1-5:

1. No issue with changing the b to B.

2. Please  don't remove the close and submit buttons as suggested. (1) On new installations there are times I close a new task without adding any settings. (2) My recent WakeOnLan plugin auto-populates device settings on a Submit, so saving an "empty" device is valid.

3.  Current behavior seems normal to me. 

4. Appears this issue is being resolved by TD-er.

5. I prefer the current behavior. Many times I have deleted a device and immediately selected a different one in its place. The requested change would add more steps in this scenario.

\-Thomas

Misiu commented:
@thomastech ad 2 I was thinking about this place:
![image](https://user-images.githubusercontent.com/1741838/98964627-58390f80-2509-11eb-87eb-cb910ec311e9.png)
after selecting a value from the dropdown you instantly go to the device page, but on that page are two buttons, not Submit is useless.
I **want that button to stay**, I suggested removing the auto-submit function by removing `onchange="return dept_onchange(frmselect)"`

Rest was just a suggestion 🙂

thomastech commented:
@Misiu, I'm not bothered by the currently displayed button choices or the auto-submit. I'll admit the submit button seems unnecessary in that screenshot, but it is comforting to see it; It provides a consistent presentation throughout the device pages. And the auto-submit hasn't been an issue for me.

Long story short, I'll let you and TD-er work it out.

\- Thomas

uzi18 commented:
@TD-er thx for clean up, it is a massive change in code, and hard to find now something erroneous.
In real-world it is harder to work, that's why we did not merge 4 sensors per task.
If there are one or more sensors on the bus (gpio) connected in parasite mode you can't send 1-4 conversion commands to all of them. 
Simply because every conversion command (bus reset in fact) will reset all of parasite powered sensors so the command will be ineffective for rest of sensors (they will start with 85deg.)
We should track parasite power status, and act according to situation:
* no parasites - work like you introduced (or send "convert all" command)
* more than 1 parasite - leave bus in high state if possible, send convert command to nonparasites like you introduced, but for parasites send convert, wait, fetch sequentially one by one.

So in worst case we will block gpio for almost 4s. on one task.
Did you change something in timmings or more like only check if they are accurate?

Fake sensors will show as parasites (they need more power than original dallases).

Just my 50c.

TD-er commented:
I did change the timings as they were not that constant.
Also the short pulses could be quite short occasionally.
It is now based on the usec timer and is now a "busy wait" for the short intervals as a delaymicros often is only the minimal time set with quite some jitter.

I am now looking into the parasitic powered sensors as how to use them.
When the sensors are powered via the Vdd pin you can still communicate to the bus to address others.
But as soon as you have parasite powered sensors, you must keep the data pin high during the measurement. So the sensors cannot be read in parallel, as you also state.
I plan on only allow parasite powered sensors with 1 sensor per task. Multiple sensors per task only when none of those are powered via the data line
And also the other 2 Dallas plugins must be adapted to be used with Dallas temp sensors
For example iButton should not be used on the same pin as parasitic powered temp sensors as those block the pin for upto 800 msec so you cannot read an iButton during reading a temp sensor. (in parasitic mode)
It should not be a problem when having powered temp sensors on the same GPIO

Screenshot of the current state of this PR (including last commit)
![image](https://user-images.githubusercontent.com/3751318/98970618-6e969980-2510-11eb-9ed4-3f7ec88f28f1.png)

uzi18 commented:
@TD-er just want to mention - timings were adjusted to tolerate longer wires/more sensors (capacitance) on bus.
You can try a different sets and chack if failed samples decrease.

TD-er commented:
The short pulses as they were (some times) would not have worked well on longer cables, or cables with lots of sensors.
They simply were every now and then too short.
Now the tolerances of the timings do seem to be less or equal to that of the sensors I tested. (higher tolerances of the sensors themselves in practice compared to what the datasheet mentions)

TD-er committed:
[Cleanup] Rename _CPlugin_SensorTypeHelper.h to _Plugin_SensorTypeHelper.h

It is meant for plugins, not controllers, so using the CPlugin prefix was confusing.

TD-er committed:
[Cleanup] Load default var names when changing nr output values

TD-er committed:
[LoRaWAN] C018 Reduce memory usage when controller not used

TD-er committed:
[Cleanup] Save resources by moving static strings to flash

TD-er committed:
[GPS] Read all data available in the serial buffer

Misiu commented:
@TD-er I have both normal and parasites DS18B20 (both genuine and clones) I can test them with new builds if you want

TD-er commented:
@Misiu 
I will soon merge this PR and implement the parasite support later.
In my test setup here, the Dallas support works exceptionally well on my breadboard (6 sensors with pull-up resistor and powered at 3v3) with less than 1 promille of the samples returning a NaN (or whatever is set as error state value)

I will make a test build with the current state for you, it would be great to see if something is broken.
I don't have the Dallas counter nor an iButton, so I cannot test those to see if they still work.

TD-er commented:
Made a new test build:
-[ESPEasy_ESP32_mega-20201102-33-PR_3367.zip](https://www.dropbox.com/s/cu0l3qevmt0g659/ESPEasy_ESP32_mega-20201102-33-PR_3367.zip?dl=0)
-[ESPEasy_ESP82xx_mega-20201102-33-PR_3367.zip](https://www.dropbox.com/s/zb4z8bicas7yght/ESPEasy_ESP82xx_mega-20201102-33-PR_3367.zip?dl=0)

Included are some optimizations to make it easier to setup and also some cleanup of strings I found that were not marked as flash strings.

Misiu commented:
@TD-er I'll try to test the build today.
I have RW1990(https://www.ibutton.cc/products/ibutton/rw1990/), which is a read-write version of iButton, so I can test it 🙂

P.S.
how are you connecting iButton? Are you using only a single 4,7k resistor?

TD-er committed:
[Dallas] Document + update TTN decoder

thomastech commented:
**Warning:** Mcafee reports that the ardublock.ru link in the previous post is infected (trojan). I'd advise removing the link. Plus scan your PC for virus if you've already clicked it.

\- Thomas

Misiu commented:
> **Warning:** Mcafee reports that the ardublock.ru link in the previous post is infected (trojan). I'd advise removing the link. Plus scan your PC for virus if you've already clicked it.
> 
> - Thomas

I've edited my previous comment. On my PC (I have NOD32) I had no problems.

Misiu commented:
@TD-er what build has P080 (iButton) included? I'd like to test it.

TD-er commented:
> @TD-er what build has P080 (iButton) included? I'd like to test it.

Should be any "testing" build.

tonhuisman commented:
P080 is included in the TESTING (test) builds

TD-er committed:
Fix build issue on ESP32

TD-er committed:
Merge branch 'mega' into bugfix/dallas_multiple_instances

TD-er committed:
Include all Dallas plugins in the Custom builds

Misiu commented:
@TD-er  @tonhuisman thanks, now I have that plugin on the list.
This works, kind of.
What I mean by that:
1. When I have connected iButton I can add/edit the plugin instance (Task), but when iButton is disconnected and I go to the Edit Task page the device address is empty. This way you won't be able to do any changes to Task, for example, change the interval.
You must plug the Button or put it on the reader.
2. After checking send to Controller (Home Assistant (openHAB) MQTT) I don't see any values on the MQTT Broker (DS18B20 works fine). I thin it is the same issue as described here: https://github.com/letscontrolit/ESPEasy/issues/2070
3. I have only one iButton, but I think it will be hard to add more iButtons into the same task (use case: open the gate with 4 iButtons)
![image](https://user-images.githubusercontent.com/1741838/99298375-eaaf1b00-2849-11eb-8e86-da603d40f67e.png)

iButton disconnected:
![image](https://user-images.githubusercontent.com/1741838/99299416-7b3a2b00-284b-11eb-97a7-cd7f4d4ce135.png)

I don't want to write here about iButton, so if you want I can comment in the linked issue.

TD-er committed:
[Servo] Add normal build for ESP32 including servo support

TD-er commented:
I think I know what you mean here.
I have ordered a few iButton things so I can test myself, but they have not arrived yet. (expect them in 2 weeks)

The current implementation of the iButton task has not changed, so it was already "impossible to use".
As soon as I have the units, I will make it "well usable".

TD-er committed:
[Servo] Fix lib_ignore for all ESP82xx builds to ignore ServoESP32

Misiu commented:
@TD-er sure thing 🙂
I just wanted to give initial feedback about iButton and potential usage.
Probably Plugin is rescanning for iButtons when you enter Task Edit Page, so it won't be possible to assign more than one iButton to a single switch. I think it will be a valid use case - to give iButton to each family member. So there should be an option to select the new ID and add it to the list of valid IDs and also a way to manage the list (remove the ID from the list).
I also think that iButton Plugin should send iButton ID (address) via MQTT (to the controller). This way we could write the logic in openHab, Home Assistant, or in other systems.

BTW order RW1990, they are almost identical to DS1990, but you can programmatically assign an ID to them. Really nice feature 😁

TD-er commented:
> BTW order RW1990, they are almost identical to DS1990, but you can programmatically assign an ID to them. Really nice feature 😁

Yep did that :)

And I do have some ideas for the iButton modules.
For example they should be added to a list of modules, along with a shorter "group".
This way you can just generate an event with the scanned tag and group as eventvalues.
Using group tags would make processing in rules a lot easier to do and you can just update a tag list if a tag should be assigned to a specific group (or removed)

KoljaW commented:
I have successfully connected 4 iButtons and checked the box "Send to Controller" but i can't see any data in ioBroker. iButton support would be amazing. I hope it will be implemented soon :)

TD-er commented:
> I have successfully connected 4 iButtons and checked the box "Send to Controller" but i can't see any data in ioBroker. iButton support would be amazing. I hope it will be implemented soon :)

I hope so too, or else I threw away money on these :)
![image](https://user-images.githubusercontent.com/3751318/102278526-47295700-3f2a-11eb-8704-c9edffe2e3fc.png)

KoljaW commented:
Looks good. i am counting on you because my skills are not good enough for fixing the mqtt implementation.

KoljaW commented:
Any news? Btw happy new year.

TD-er commented:
Not yet, as I have been working on the #3439 PR and now I will work (as promised in another issue) on the 1M ESP units with problematic WiFi.
But after that, the bag of Dallas keys is the next on my desk.
Let's hope things will not fall of my desk or else historic order will be lost ;)

Misiu commented:
> > BTW order RW1990, they are almost identical to DS1990, but you can programmatically assign an ID to them. Really nice feature 😁
> 
> Yep did that :)

@TD-er any updates on the iButton plugin?
Also, did you try to write the serial number to RW1990? I ordered 10 RW1990 and I can only program 4 out of 10. What is the success rate with your batch?

TD-er commented:
Nope not yet.
All time went into WiFi related issues lately.

KoljaW commented:
@TD-er A new schedule for iButtons implementation? Thanks

TD-er commented:
Nope, not at this moment.
I've been way too busy with other parts of the code, like reported issues and the ESPEasy-now mesh and other activities.
